src/main/java/org/broadinstitute/hellbender/tools/spark/sv/FindBreakpointEvidenceSpark.java

```diff
@@ -1329,22 +1273,65 @@ public KmerAndCount read(final Kryo kryo, final Input input, final Class<KmerAnd
             this.kmerMultiMap = kmerMultiMap;
         }
 
-        public Iterator<QNameAndInterval> apply(final GATKRead read) {
-            intervalIdSet.clear();
+        public Iterator<Tuple2<SVKmer, String>> apply( final GATKRead read ) {
+            List<Tuple2<SVKmer, String>> results = new ArrayList<>();
             SVKmerizerWithLowComplexityFilter.stream(read.getBases(), kSize, minEntropy)
                     .map( kmer -> kmer.canonical(kSize) )
                     .forEach( kmer -> {
                         final Iterator<KmerAndInterval> itr = kmerMultiMap.findEach(kmer);
-                        while ( itr.hasNext() ) {
-                            intervalIdSet.add(itr.next().getValue());
-                        }
+                        if ( itr.hasNext() ) results.add(new Tuple2<>(kmer, read.getName()));
                     });
-            if (intervalIdSet.isEmpty()) return emptyIterator;
-            qNameAndIntervalList.clear();
-            final String qName = read.getName();
-            intervalIdSet.forEach(intervalId ->
-                    qNameAndIntervalList.add(new QNameAndInterval(qName, intervalId)));
-            return qNameAndIntervalList.iterator();
+            return results.iterator();
+        }
+    }
+
+    private static final class KmerQNameToQNameIntervalMapper {
+        private final HopscotchUniqueMultiMap<SVKmer, Integer, KmerAndInterval> kmerMultiMap;
+        private final int maxQNamesPerKmer;
+        private final int kmerMapSize;
+
+        KmerQNameToQNameIntervalMapper( final HopscotchUniqueMultiMap<SVKmer, Integer, KmerAndInterval> kmerMultiMap,
+                                        final int maxQNamesPerKmer,
+                                        final int kmerMapSize ) {
+            this.kmerMultiMap = kmerMultiMap;
+            this.maxQNamesPerKmer = maxQNamesPerKmer;
+            this.kmerMapSize = kmerMapSize;
+        }
+
+        public Iterable<QNameAndInterval> call( final Iterator<Tuple2<SVKmer, String>> pairItr ) {
+            HopscotchMap<SVKmer, List<String>, Map.Entry<SVKmer, List<String>>> kmerQNamesMap =
+                    new HopscotchMap<>(kmerMapSize);
+            while ( pairItr.hasNext() ) {
+                final Tuple2<SVKmer, String> pair = pairItr.next();
+                final SVKmer kmer = pair._1();
+                Map.Entry<SVKmer, List<String>> entry = kmerQNamesMap.find(kmer);
+                if ( entry == null ) {
+                    entry = new AbstractMap.SimpleEntry<>(kmer, new ArrayList<>());
+                    kmerQNamesMap.add(entry);
+                }
+                List<String> qNames = entry.getValue();
+                if ( qNames != null ) {
```

Comment:
Maybe a comment upfront explaining that nulls in the values mean that the max qnames per kmer has been exceeded would help slow folks like me understand (since the null checks looks like the sort of boilerplate one might see in an iteration that populates a map).